syntax = "proto3";
package pbfProto;

// User
enum ROLE {
    FEMTECH_USER = 0;
    FEMTECH_PROVIDER = 1;
    REGULATORY_AUTHORITY = 2;
    THIRD_PARTY = 3;
    RESEARCH_INSTITUTE = 4;
}

message UserAccountData {
    string profile_name = 1;
    string registration_date = 2;
    string payment_plan = 3;
    bool subscription_status = 4;
}

message UserProfileData {
    string age = 1; 
    string menstrualCycleLength = 2;
    string periodLength = 3;
    string mainContraceptiveMethod = 4; 
}

message User {
    ROLE role = 1;
    string registration_date = 2;
    string profile_name = 3;
    string username = 4;
}

message UserRegistrationRequest {
    UserAccountData accountData = 1;
    string username = 2;
    string password = 3;
    ROLE role = 4;
    UserProfileData profileData = 5;
}

message ErrorMessage {
    string description = 1;
}

message UserRegistrationResponse {
    ErrorMessage errorMessage = 1;
    User user = 2;
}

message GetAccountDataRequest {
    string username = 1;
}

message GetAccountDataResponse {
    UserAccountData accountData = 1;
    ErrorMessage errorMessage = 2;
}

message GetProfileDataRequest {
    string username = 1;
}

message GetProfileDataResponse {
    UserProfileData profileData = 1;
    ErrorMessage errorMessage = 2;
}


message UserAuthorizationRequest {
    string token = 1;
}

message UserAuthorizationResponse {
    ROLE role = 1;
    string username = 2;
    ErrorMessage errorMessage = 3;
}

message UserAuthenticationRequest {
    string username = 1;
    string password = 2;
}

message UserAuthenticationResponse {
    string token = 1;
    ErrorMessage errorMessage = 2;
}

message UserRoleAuthorizationRequest {
    string token = 1;
}

message UserRoleAuthorizationResponse {
    ROLE role = 1;
    ErrorMessage errorMessage = 2;
}

message UserIdAuthorizationRequest {
    string token = 1;
}

message UserIdAuthorizationResponse {
    string id = 1;
    ErrorMessage errorMessage = 2;
}

message UserLogoutRequest {
    string token = 1;
}

message UserLogoutResponse {
    ErrorMessage errorMessage = 1;
}

message UserDeleteRequest {
    string token = 1;
}

message UserDeleteResponse {
    ErrorMessage errorMessage = 1;
}


